nomos-core/chain-defs/src/proofs/leader_proof.rs

```diff
@@ -1,100 +1,379 @@
+        #[cfg(feature = "pol-dev-mode")]
+        if std::env::var(POL_PROOF_DEV_MODE).is_ok() {
+            tracing::warn!(
+                "Proofs are being verified in dev mode. This should never be used in production."
+            );
+            return &self.public == public_inputs;
+        }
+
+        let leader_pk = ed25519_pk_to_fr_tuple(self.leader_key());
+        pol::verify(
+            &self.proof,
+            &pol::PolVerifierInput::new(
+                self.entropy(),
+                public_inputs.slot,
+                public_inputs.epoch_nonce,
+                public_inputs.aged_root,
+                public_inputs.latest_root,
+                public_inputs.total_stake,
+                leader_pk,
+            ),
+        )
+        .is_ok()
+    }
+
+    fn entropy(&self) -> Fr {
+        self.entropy_contribution
+    }
+
+    fn leader_key(&self) -> &ed25519_dalek::VerifyingKey {
+        &self.leader_key
+    }
+
+    fn voucher_cm(&self) -> &VoucherCm {
+        &self.voucher_cm
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LeaderPublic {
+    pub slot: u64,
+    #[serde(with = "serde_fr")]
+    pub epoch_nonce: Fr,
+    pub total_stake: u64,
+    #[serde(with = "serde_fr")]
+    pub aged_root: Fr,
+    #[serde(with = "serde_fr")]
+    pub latest_root: Fr,
+}
+
+impl LeaderPublic {
+    #[must_use]
+    pub const fn new(
+        aged_root: Fr,
+        latest_root: Fr,
+        epoch_nonce: Fr,
+        slot: u64,
+        total_stake: u64,
+    ) -> Self {
+        Self {
+            slot,
+            epoch_nonce,
+            total_stake,
+            aged_root,
+            latest_root,
+        }
+    }
+
+    #[must_use]
+    pub fn check_winning(&self, value: u64, note_id: Fr, sk: Fr) -> bool {
+        let (t0, t1) = self.scaled_phi_approx();
+        let threshold =
+            Self::phi_approx(&Fr::from(value), &(Fr::from(t0), Fr::from(t1))).into_bigint();
+        let ticket = Self::ticket(note_id, sk, self.epoch_nonce, Fr::from(self.slot)).into_bigint();
+        ticket < threshold
+    }
+
+    #[must_use]
+    #[cfg(feature = "pol-dev-mode")]
+    pub fn check_winning_dev(
+        &self,
+        value: u64,
+        note_id: Fr,
+        sk: Fr,
+        active_slot_coeff: f64,
+    ) -> bool {
+        let (t0, t1) = self.scaled_phi_approx_dev(active_slot_coeff);
+        let threshold =
+            Self::phi_approx(&Fr::from(value), &(Fr::from(t0), Fr::from(t1))).into_bigint();
+        let ticket = Self::ticket(note_id, sk, self.epoch_nonce, Fr::from(self.slot)).into_bigint();
+        ticket < threshold
+    }
+
+    fn scaled_phi_approx(&self) -> (BigUint, BigUint) {
+        let t0 = &*pol::T0_CONSTANT / &BigUint::from(self.total_stake);
+        let total_stake_sq = &BigUint::from(self.total_stake) * &BigUint::from(self.total_stake);
+        let t1 = &*pol::P - (&*pol::T1_CONSTANT / &total_stake_sq);
+        (t0, t1)
+    }
```

Comment:
Are you missing a factor of 1 / 2 in t1?
```suggestion
    fn scaled_phi_approx(&self) -> (BigUint, BigUint) {
        let t0 = &*pol::T0_CONSTANT / &BigUint::from(self.total_stake);
        let total_stake_sq = &BigUint::from(self.total_stake) * &BigUint::from(self.total_stake);
        let t1 = &*pol::P - (&*pol::T1_CONSTANT / (2u64 * &total_stake_sq));
        (t0, t1)
    }
```